File: scripts/gce/docker/caching-experiment.pl
#!/usr/bin/perl

use warnings;
use strict;

my $usage = "usage: ./caching-experiment.pl user image";
die $usage unless @ARGV == 2;

my ($user, $image) = @ARGV;

my $GCE_IP = "104.196.38.63";
my $USE_REDIS = 1;

my $outputDir = "../../experiments/results/caching";
my $gceOutputDir = "caching";

# set up log
my $log = "caching-log.txt";
system("rm -f $log; touch $log");

my $startDiamondCmd = "ssh -t $GCE_IP 'cd diamond-src/scripts; ./manage-servers.py start ../platform/test/gcelocaloneshard' >> $log 2>&1";
my $killDiamondCmd = "ssh $GCE_IP 'cd diamond-src/scripts; ./manage-servers.py kill ../platform/test/gcelocaloneshard' >> $log 2>&1";
my $startRedisCmd = "ssh -f $GCE_IP 'nohup redis-3.0.7/src/redis-server &' >> $log 2>&1";
my $killRedisCmd = "ssh $GCE_IP 'pkill redis-server'";

# make sure Diamond and redis servers are not running
system("$killRedisCmd");
system("$killDiamondCmd");

# do sanity checks
system("$startRedisCmd");
my $checkResult = system("./sanity-checks.pl run_baseline.py $outputDir $gceOutputDir $GCE_IP $USE_REDIS >> $log 2>&1");
if ($checkResult != 0) {
    die("Error in sanity checks: see log file for details");
}
system("$killRedisCmd");

# run experiments
runDiamond("caching", 200, [2, 4, 6, 8, 10]);
runDiamond("nocaching", 200, [2, 4, 6, 8, 10]);



sub runDiamond {
    my ($caching, $numPairs, $instanceNumsRef) = @_;
    my @instanceNums = @{$instanceNumsRef};

    # reset client VM
    logPrint("Running notification experiment with caching $caching\n");

    my $outFile = "$outputDir/diamond.$caching.txt";
    open(OUTFILE, "> $outFile");
    print(OUTFILE "clients\tthroughput\tlatency\tseconds\tinstances\n");
    for my $instances (@instanceNums) {
        logPrint("Running $instances instances...\n");

        # start redis and Diamond servers
        resetClient();
        system("$startRedisCmd");
        system("$startDiamondCmd");

        # clear output location
        if ($USE_REDIS) {
            system("ssh $GCE_IP 'redis-3.0.7/src/redis-cli -p 6379 flushdb' >> $log 2>&1");
        }
        else {
            system("ssh $GCE_IP 'rm diamond-src/scripts/experiments/$gceOutputDir/*' >> $log 2>&1");
        }

        # run Kubernetes instances
        my $cachingArg = "";
        if ($caching eq "nocaching") {
            $cachingArg = " --nocaching ";
        }
        system("./run-kubernetes-job.pl caching $image $user $instances run_game.py --config gcelocaloneshard --numpairs $numPairs $cachingArg >> $log 2>&1");

        # parse output
        my $clientCmd = "ls diamond-src/scripts/experiments/baseline | wc | awk \"{ print \\\$1 }\"";
        my $resultCmd = "cd diamond-src/scripts/experiments; ./parse-scalability.py -d $gceOutputDir";
        if ($USE_REDIS) {
            $clientCmd = "redis-3.0.7/src/redis-cli -p 6379 get clients";
            $resultCmd = "diamond-src/scripts/experiments/parse-game.py -r";
        }
        my $clients = `ssh $GCE_IP '$clientCmd'`;
        chomp($clients);
        my @result = `ssh $GCE_IP '$resultCmd'`;

        my $throughput = "ERROR";
        my $latency = "ERROR";
        my $abortrate = "ERROR";
        my $seconds = "ERROR";
        for my $line (@result) {
            if ($line =~ /^Avg\. turns\/second: ([\d\.]+)$/) {
                $throughput = $1;
            }
            elsif ($line =~ /^Avg\. time between turns \(s\): ([\d\.]+)$/) {
                $latency = $1;
            }
            elsif ($line =~ /over ([\d\.]+) seconds/) {
                $seconds = $1;
            }
        }
        print(OUTFILE "$clients\t$throughput\t$latency\t$seconds\t$instances\n");

        # kill redis and Diamond servers
        system("$killRedisCmd");
        system("$killDiamondCmd");
    }
    close(OUTFILE);
    my $time = time();
    system("cp $outFile $outFile.$time");

    return;
}

sub logPrint {
    my ($str) = @_;
    chomp($str);
    system("echo $str | tee -a $log");
    return;
}

sub resetClient {
    system("gcloud compute instances reset diamond-client --zone us-east1-c >> $log 2>&1");
    my $done = 0;
    while (!$done) {
        sleep(5);
        my $test = `ssh $GCE_IP 'echo test' 2>&1`;
        if ($test =~ /test/) {
            $done = 1;
        }
        elsif ($test =~ /Connection refused/) {
            logPrint("Waiting for GCE client VM to restart...");
        }
        else {
            logPrint("Error: unknown response to GCE client VM check: $test");
            exit(1);
        }
    }
}

File: scripts/experiments/caching-experiment.py
#!/usr/bin/python

import experiment_common
from experiment_common import logPrint
import re
import subprocess
import sys

CONFIG_PREFIX = "diamond-src/platform/test/niel"
BATCH_SIZE = 64

OUTPUT_DIR = "results/caching"
LOG = "caching-log.txt"

SRC_HOST = experiment_common.getSrcHost()

experiment_common.setLog(LOG)
experiment_common.init()

experiment_common.copyFromSrcHostToClients("diamond-src/scripts/experiments/run_game.py")
experiment_common.copyFromSrcHostToClients("diamond-src/scripts/experiments/run_redis_game.py")
experiment_common.copyFromSrcHostToClients("diamond-src/scripts/experiments/client_common.py")

def runBaseline(numPairsPerMachine, machineNums):
    logPrint("Running baseline")
    outFileName = OUTPUT_DIR + "/baseline.txt"
    outFile = open(outFileName, "w")

    startBaselineCmd = "ssh -t %s 'diamond-src/apps/baseline-benchmarks/100game/manage-baseline-100game-servers.py start %s' >> %s 2>&1" \
            % (SRC_HOST, CONFIG_PREFIX, LOG)
    killBaselineCmd = "ssh %s 'diamond-src/apps/baseline-benchmarks/100game/manage-baseline-100game-servers.py kill %s' >> %s 2>&1" % (SRC_HOST, CONFIG_PREFIX, LOG)

    # read hostname and port of Redis leader from config file
    hostname = ""
    port = ""
    config = subprocess.check_output("ssh %s 'cat %s0.config'" % (SRC_HOST, CONFIG_PREFIX), shell=True).split("\n")
    match = re.match("replica\s+([\d\w\.]+):(\d+)", config[1])
    if match:
        hostname = match.group(1)
        port = match.group(2)
    else:
        print("Error reading hostname and port of Redis leader from config file")
        sys.exit(1)

    experiment_common.startDataRedis()

    outFile.write("clients\tthroughput\tlatency\tseconds\tmachines\n")

    for numMachines in machineNums:
        logPrint("Running %d machines" % numMachines)
        subprocess.call(startBaselineCmd, shell=True)
        experiment_common.clearDataRedis()

        command = "./run_redis_game.py --hostname %s --port %s --numpairs %d" \
                % (hostname, port, numPairsPerMachine)
        experiment_common.runOnClientMachines(command, numMachines)

        clients = experiment_common.getClientCountFromDataRedis()
        result = subprocess.check_output("ssh %s 'diamond-src/scripts/experiments/parse-game.py -r'" % experiment_common.getSrcHost(), shell=True)
        for line in result.split("\n"):
            match = re.match("Avg\. turns\/second: ([\d\.]+)", line)
            if match:
                throughput = match.group(1)
            match = re.match("Avg\. time between turns \(s\): ([\d\.]+)", line)
            if match:
                latency = match.group(1)
            match = re.search("over ([\d\.]+) seconds", line)
            if match:
                seconds = match.group(1)
        outFile.write("%d\t%s\t%s\t%s\t%d\n" % (clients, throughput, latency, seconds, numMachines))

        subprocess.call(killBaselineCmd, shell=True)

def runDiamond(caching, numPairsPerMachine, machineNums):
    logPrint("Running Diamond with caching %s" % (caching))
    outFileName = ""
    if caching:
        outFileName = OUTPUT_DIR + "/diamond.caching.txt"
    else:
        outFileName = OUTPUT_DIR + "/diamond.nocaching.txt"
    outFile = open(outFileName, "w")

    experiment_common.startDataRedis()

    outFile.write("clients\tthroughput\tlatency\tseconds\tmachines\n")

    for numMachines in machineNums:
        logPrint("Running %d machines" % numMachines)
        experiment_common.startDiamond(CONFIG_PREFIX, batchSize=64)
        experiment_common.clearDataRedis()

        cachingArg = ""
        if not caching:
            cachingArg = "--nocaching"
        command = "./run_game.py --config %s --numpairs %d %s" % (CONFIG_PREFIX, numPairsPerMachine, cachingArg)
        experiment_common.runOnClientMachines(command, numMachines)

        clients = experiment_common.getClientCountFromDataRedis()
        result = subprocess.check_output("ssh %s 'diamond-src/scripts/experiments/parse-game.py -r'" % experiment_common.getSrcHost(), shell=True)
        for line in result.split("\n"):
            match = re.match("Avg\. turns\/second: ([\d\.]+)", line)
            if match:
                throughput = match.group(1)
            match = re.match("Avg\. time between turns \(s\): ([\d\.]+)", line)
            if match:
                latency = match.group(1)
            match = re.search("over ([\d\.]+) seconds", line)
            if match:
                seconds = match.group(1)
        outFile.write("%d\t%s\t%s\t%s\t%d\n" % (clients, throughput, latency, seconds, numMachines))

        experiment_common.killDiamond(CONFIG_PREFIX)

    outFile.close()

# run experiments
runDiamond(True, 10, [1])
runDiamond(False, 10, [1])
runBaseline(10, [1])
